Build: Output package type declarations

sirreal opened the pull request:
## Description

Output TypeScript declaration files (`.d.ts`) from package sources and expose them with published packages.

[TypeScript 3.7 added support](https://devblogs.microsoft.com/typescript/announcing-typescript-3-7/#user-content---declaration-and--allowjs
) for outputting a declaration files from JavaScript sources.

The setup is largely inspired by the [jest](https://github.com/facebook/jest) setup, which is another monorepo which has many TypeScript packages they generate types from.

- Upgrade to `typescript@3.8`.
- Add bin for typechecking changed packages.
- Remove `lint-types` script, replaced by `build:package-types`.
- Restructure TypeScript configuration:
  - One base `tsconfig.base.json` that packages extend. This is not a project, it's only for extension.
  - A root `tsconfig.json` that references all typed packages.
  - Typed packages output their types at `build-types` and expose them in the published package.
  - Use [_project references_](https://www.typescriptlang.org/docs/handbook/project-references.html) to work within the monorepo (packages reference each other).
- Fix/improve some type issues.

TypeScript declaration files will be managed via `tsc --build`, the TypeScript compiler build tool for composite projects. `tsc` will manage determine the correct order to build packages so that dependencies are satisfied and rebuild only when necessary.

Additional packages can opt-in to typing by (this is documented in a new section in `packages/README.md`):
  - Adding `tsconfig.json` in their package directory.
  - Updating their `package.json` to point to the types: `{ "types": "build-types" }`.
  - Adding a reference to the root `tsconfig.json`.

## Testing

Verify the following scripts work as expected:

- `npm run build:package-types` - Build package types. Check for a generated declaration files, e.g. `packages/a11y/build-types/*.d.ts`.
- `npm run clean:package-types` - `build-types/**/*.d.ts` and `*.tsbuildinfo` files should be removed.
- `npm run build:packages` should generate the types
- Changes to package types should result in updated declarations
- Repeated type builds should complete without error.

Generated type definitions can be inspected under `packages/*/build-types`. For example:

<details>
<summary><code>@wordpress/is-shallow-equal</code></summary>

### Generated:

`index.d.ts`

```ts
export = isShallowEqual;
/**
 * @typedef {{[key: string]: any}} ComparableObject
 */
/**
 * Returns true if the two arrays or objects are shallow equal, or false
 * otherwise.
 *
 * @param {any[]|ComparableObject} a First object or array to compare.
 * @param {any[]|ComparableObject} b Second object or array to compare.
 *
 * @return {boolean} Whether the two values are shallow equal.
 */
declare function isShallowEqual(a: any[] | {
    [key: string]: any;
}, b: any[] | {
    [key: string]: any;
}): boolean;
declare namespace isShallowEqual {
    export { isShallowEqualObjects, isShallowEqualArrays, ComparableObject };
}
/**
 * Internal dependencies;
 */
declare var isShallowEqualObjects: typeof import("./objects");
declare var isShallowEqualArrays: typeof import("./arrays");
type ComparableObject = {
    [key: string]: any;
};
//# sourceMappingURL=index.d.ts.map
```

`arrays.d.ts`

```ts
export = isShallowEqualArrays;
/**
 * Returns true if the two arrays are shallow equal, or false otherwise.
 *
 * @param {any[]} a First array to compare.
 * @param {any[]} b Second array to compare.
 *
 * @return {boolean} Whether the two arrays are shallow equal.
 */
declare function isShallowEqualArrays(a: any[], b: any[]): boolean;
//# sourceMappingURL=arrays.d.ts.map
```

`objects.d.ts`

```ts
export = isShallowEqualObjects;
/**
 * Returns true if the two objects are shallow equal, or false otherwise.
 *
 * @param {import('.').ComparableObject} a First object to compare.
 * @param {import('.').ComparableObject} b Second object to compare.
 *
 * @return {boolean} Whether the two objects are shallow equal.
 */
declare function isShallowEqualObjects(a: {
    [key: string]: any;
}, b: {
    [key: string]: any;
}): boolean;
//# sourceMappingURL=objects.d.ts.map
```

[Manual type declaration](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__is-shallow-equal/index.d.ts):

```ts
// Type definitions for @wordpress/is-shallow-equal 1.4
// Project: https://github.com/WordPress/gutenberg/tree/master/packages/is-shallow-equal/README.md
// Definitions by: Derek Sifford <https://github.com/dsifford>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
// TypeScript Version: 3.5

declare const isShallowEqual: {
    /**
     * Returns `true` if the two arrays or objects are shallow equal, or `false` otherwise.
     *
     * @param a - First object or array to compare.
     * @param b - Second object or array to compare.
     *
     * @returns Whether the two values are shallow equal.
     */
    (a: ArrayLike<any> | object, b: ArrayLike<any> | object): boolean;

    /**
     * Returns true if the two arrays are shallow equal, or false otherwise.
     *
     * @param a - First array to compare.
     * @param b - Second array to compare.
     *
     * @returns Whether the two arrays are shallow equal.
     */
    isShallowEqualArrays(a: ArrayLike<any>, b: ArrayLike<any>): boolean;

    /**
     * Returns `true` if the two objects are shallow equal, or `false` otherwise.
     *
     * @param a - First object to compare.
     * @param b - Second object to compare.
     *
     * @returns Whether the two objects are shallow equal.
     */
    isShallowEqualObjects(a: object, b: object): boolean;
};

export = isShallowEqual;
```

</details>

<details>
<summary><code>@wordpress/i18n</code></summary>

### Generated

`index.d.ts`

```ts
/**
 * Merges locale data into the Tannin instance by domain. Accepts data in a
 * Jed-formatted JSON object shape.
 *
 * @see http://messageformat.github.io/Jed/
 *
 * @param {LocaleData} [data]   Locale data configuration.
 * @param {string}     [domain] Domain for which configuration applies.
 */
export function setLocaleData(data?: {
    [key: string]: any;
} | undefined, domain?: string | undefined): void;
/**
 * Retrieve the translation of text.
 *
 * @see https://developer.wordpress.org/reference/functions/__/
 *
 * @param {string} text     Text to translate.
 * @param {string} [domain] Domain to retrieve the translated text.
 *
 * @return {string} Translated text.
 */
export function __(text: string, domain?: string | undefined): string;
/**
 * Retrieve translated string with gettext context.
 *
 * @see https://developer.wordpress.org/reference/functions/_x/
 *
 * @param {string} text     Text to translate.
 * @param {string} context  Context information for the translators.
 * @param {string} [domain] Domain to retrieve the translated text.
 *
 * @return {string} Translated context string without pipe.
 */
export function _x(text: string, context: string, domain?: string | undefined): string;
/**
 * Translates and retrieves the singular or plural form based on the supplied
 * number.
 *
 * @see https://developer.wordpress.org/reference/functions/_n/
 *
 * @param {string} single   The text to be used if the number is singular.
 * @param {string} plural   The text to be used if the number is plural.
 * @param {number} number   The number to compare against to use either the
 *                          singular or plural form.
 * @param {string} [domain] Domain to retrieve the translated text.
 *
 * @return {string} The translated singular or plural form.
 */
export function _n(single: string, plural: string, number: number, domain?: string | undefined): string;
/**
 * Translates and retrieves the singular or plural form based on the supplied
 * number, with gettext context.
 *
 * @see https://developer.wordpress.org/reference/functions/_nx/
 *
 * @param {string} single   The text to be used if the number is singular.
 * @param {string} plural   The text to be used if the number is plural.
 * @param {number} number   The number to compare against to use either the
 *                          singular or plural form.
 * @param {string} context  Context information for the translators.
 * @param {string} [domain] Domain to retrieve the translated text.
 *
 * @return {string} The translated singular or plural form.
 */
export function _nx(single: string, plural: string, number: number, context: string, domain?: string | undefined): string;
/**
 * Check if current locale is RTL.
 *
 * **RTL (Right To Left)** is a locale property indicating that text is written from right to left.
 * For example, the `he` locale (for Hebrew) specifies right-to-left. Arabic (ar) is another common
 * language written RTL. The opposite of RTL, LTR (Left To Right) is used in other languages,
 * including English (`en`, `en-US`, `en-GB`, etc.), Spanish (`es`), and French (`fr`).
 *
 * @return {boolean} Whether locale is RTL.
 */
export function isRTL(): boolean;
/**
 * Returns a formatted string. If an error occurs in applying the format, the
 * original format string is returned.
 *
 * @param {string}    format The format of the string to generate.
 * @param {...string} args   Arguments to apply to the format.
 *
 * @see http://www.diveintojavascript.com/projects/javascript-sprintf
 *
 * @return {string} The formatted string.
 */
export function sprintf(format: string, ...args: string[]): string;
export type LocaleData = {
    [key: string]: any;
};
//# sourceMappingURL=index.d.ts.map
```

### [Manual](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__i18n/index.d.ts)

```ts
// Type definitions for @wordpress/i18n 3.4
// Project: https://github.com/WordPress/gutenberg/tree/master/packages/i18n/README.md
// Definitions by: Derek Sifford <https://github.com/dsifford>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

/**
 * Merges locale data into the Tannin instance by domain. Accepts data in a
 * Jed-formatted JSON object shape.
 *
 * @see http://messageformat.github.io/Jed/
 *
 * @param data   Locale data configuration.
 * @param domain Domain for which configuration applies.
 */
export function setLocaleData(data: any, domain?: string): void;

/**
 * Retrieve the translation of text.
 *
 * @see https://developer.wordpress.org/reference/functions/__/
 *
 * @param  text   Text to translate.
 * @param  domain Domain to retrieve the translated text.
 */
export function __(text: string, domain?: string): string;

/**
 * Retrieve translated string with gettext context.
 *
 * @see https://developer.wordpress.org/reference/functions/_x/
 *
 * @param  text    Text to translate.
 * @param  context Context information for the translators.
 * @param  domain  Domain to retrieve the translated text.
 */
export function _x(text: string, context: string, domain?: string): string;

/**
 * Translates and retrieves the singular or plural form based on the supplied
 * number.
 *
 * @see https://developer.wordpress.org/reference/functions/_n/
 *
 * @param  single The text to be used if the number is singular.
 * @param  plural The text to be used if the number is plural.
 * @param  number The number to compare against to use either the
 *                         singular or plural form.
 * @param  domain Domain to retrieve the translated text.
 */
export function _n(
    single: string,
    plural: string,
    n: number,
    domain?: string
): string;

/**
 * Translates and retrieves the singular or plural form based on the supplied
 * number, with gettext context.
 *
 * @see https://developer.wordpress.org/reference/functions/_nx/
 *
 * @param  single  The text to be used if the number is singular.
 * @param  plural  The text to be used if the number is plural.
 * @param  number  The number to compare against to use either the
 *                          singular or plural form.
 * @param  context Context information for the translators.
 * @param  domain  Domain to retrieve the translated text.
 */
export function _nx(
    single: string,
    plural: string,
    n: number,
    context: string,
    domain?: string
): string;

/**
 * Returns a formatted string. If an error occurs in applying the format, the
 * original format string is returned.
 *
 * @param  format  The format of the string to generate.
 * @param  args Arguments to apply to the format.
 *
 * @see http://www.diveintojavascript.com/projects/javascript-sprintf
 */
export function sprintf(format: string, ...args: any[]): string;
```

</details>

Try using some of the typed packages in an existing project, preferably via TypeScript. Are the type definitions correctly exposed?

sirreal committed:
Add tsconfigs for packages

sirreal committed:
Update TypeScript package

sirreal committed:
Add types

sirreal committed:
Gitignore types output

sirreal committed:
Clean build-types

sirreal committed:
Add missing Parameters utility type

sirreal committed:
Fix PriorityQueue types

sirreal committed:
Add sprintf-js types

sirreal committed:
Type plural_forms function

sirreal committed:
Allow i18n implicit any for untyped dependencies

sirreal committed:
Add include

sirreal committed:
Add tsbuildinfo locations

sirreal committed:
Clean the right directories

sirreal committed:
Remove root tsconfig

sirreal committed:
Add build-types

sirreal committed:
Disable token-list types

sirreal committed:
Remove ts cache from cleanup

sirreal committed:
Include build-types in files

sirreal committed:
Use "build" over "b"

sirreal commented:
I've reworked this PR and I'd appreciate feedback at this time.

gziolo commented:
https://travis-ci.com/WordPress/gutenberg/jobs/291840825#L372

> ERR! missing script: lint-types

It looks like all references need to be removed.

sirreal committed:
Remove lint-types from lint

This is now handled as part of build:package-types

sirreal commented:
> https://travis-ci.com/WordPress/gutenberg/jobs/291840825#L372
> 
> > ERR! missing script: lint-types
> 
> It looks like all references need to be removed.

I've removed the reference. Rather than linting types independently, valid types become part of the package build.

sirreal commented on package.json:
```diff
@@ -164,17 +168,18 @@
 		"build:packages": "node ./bin/packages/build.js",
-		"build": "npm run build:packages && wp-scripts build",
+		"build:package-types": "node ./bin/packages/build-types.js",
+		"build": "npm run build:packages && npm run build:package-types && wp-scripts build",
```
I'm unsure where the types build should go.
- Part of `build:packages`?
- Concurrent/before/after?

ockham replied:
No clear preference. Part of `build:packages` might make sense, so they're seen as part of the package build, rather than something extraneous (that might fall through the cracks?).

(But I think it's also fine as-is, otherwise.)

aduth replied:
At least based on the current [packages release workflow](https://github.com/WordPress/gutenberg/blob/master/packages/README.md#releasing-packages), if these are to be included in the Lerna publish, it should happen as part of `build:packages`:

https://github.com/WordPress/gutenberg/blob/8a94254b6adac030f5a6ffd8e718972e1f19f48e/package.json#L204-L206

Otherwise, those `publish:*` commands would need to be updated as well. But I think it makes sense to consider this as part of `build:packages`. The `wp-scripts` portion of the build is fine to consider as separate from the packages themselves.

sirreal replied:
I've moved it into build:packages (first in `&&` chain).

sirreal commented on packages/token-list/_tsconfig.json:
```diff
@@ -0,0 +1,9 @@
+{
+	// Currently disabled, package does not typecheck
```
Token list had it's types checked before, but doesn't type check now. Worth checking what's changed in the config in this PR. I did use a more strict set of rules.

aduth replied:
Not sure if the plan is to sort this out before merge, but I'd rather remove the type-checking altogether than to include the temporary TODO file.

aduth replied:
> I did use a more strict set of rules.

Noticing specifically a change in `noImplicitAny`, which I recall to have been the subject of previous discussion at https://github.com/WordPress/gutenberg/pull/17014#discussion_r317923126 .

ockham commented on bin/packages/build-types.js:
```diff
@@ -0,0 +1,38 @@
+		cwd: path.resolve( __dirname, '..', '..' ),
+	} );
+	const projectPaths = packagesWithTs.map( ( tsconfigPath ) =>
+		path.resolve( path.dirname( tsconfigPath ) )
```
Is the extra `resolve` needed? Shouldn't `dirname` do the right thing already?

```suggestion
		path.dirname( tsconfigPath )
```

ockham commented:
This is overall looking great -- changes make sense, and `npm run build:package-types` produces a nice set of type defs!

I noticed one problem, I ran `npm run build` after `npm run build:package-types` (out of curiosity), and got

```
> node ./bin/packages/build-types.js

packages/a11y/src/index.js:4:22 - error TS7016: Could not find a declaration file for module '@wordpress/dom-ready'. '/home/bernie/src/gutenberg/packages/dom-ready/build/index.js' implicitly has an 'any' type.
  Try `npm install @types/wordpress__dom-ready` if it exists or add a new declaration (.d.ts) file containing `declare module '@wordpress/dom-ready';`

4 import domReady from '@wordpress/dom-ready';
                       ~~~~~~~~~~~~~~~~~~~~~~
```

Every subsequent call of `npm run build:packages` also failed, even after a `npm run clean:packages`. Do we need to clean TS caches? :thinking:

sirreal committed:
Clean caches, some config cleanup

sirreal committed:
Remove redundant resolve

sirreal commented:
Rebased. 

> Every subsequent call of `npm run build:packages` also failed, even after a `npm run clean:packages`. Do we need to clean TS caches? 🤔

Yes we do, if we're cleaning the types and expect them to be reproduced, we should clean the caches, otherwise the cache tells `tsc` that the types are built, but when another package goes to find the type definitions, they're not present!

I've pushed a fix for this 👍

ockham commented:
> Rebased.
> 
> > Every subsequent call of `npm run build:packages` also failed, even after a `npm run clean:packages`. Do we need to clean TS caches? thinking
> 
> Yes we do, if we're cleaning the types and expect them to be reproduced, we should clean the caches, otherwise the cache tells `tsc` that the types are built, but when another package goes to find the type definitions, they're not present!
> 
> I've pushed a fix for this

Thanks! This is working now :+1:

ockham reviewed (approved):
This is looking good to me, and testing well!

Excited to see type declarations generated directly from Gutenberg's source :+1:

aduth commented on bin/packages/build-types.js:
```diff
@@ -0,0 +1,38 @@
+/* eslint-disable eslint-comments/disable-enable-pair, no-console */
```
I was under the impression we had some rule in place to enforce that all `eslint-disable` are complimented by an `eslint-enable`, even for these sorts of cases where we intend for the whole file to be disabled.

e.g.

https://github.com/WordPress/gutenberg/blob/34ed2a6042d42fa18a5dcd0853d59bdff6a068d9/bin/packages/build.js#L1

https://github.com/WordPress/gutenberg/blob/34ed2a6042d42fa18a5dcd0853d59bdff6a068d9/bin/packages/build.js#L184

Maybe some issue when disabling multiple in one go? Not really an issue for this pull request specifically, just wondering if it's something we should be aware of.

sirreal replied:
> I was under the impression we had some rule in place to enforce that all `eslint-disable` are complimented by an `eslint-enable`, even for these sorts of cases where we intend for the whole file to be disabled.
> Maybe some issue when disabling multiple in one go?

The other rule disabled in the multiple disable is what disabled the lint rule enforcing dis/enable pairs 😂 `eslint-comments/disable-enable-pair`.

🤔 I could add the enable at the end of the file to ensure no other less desirable dis/enable pairs sneak into this file. It might be worth overriding the no-console rule in this directory as well.

aduth commented on bin/packages/build-types.js:
```diff
@@ -0,0 +1,38 @@
+	'node_modules',
+	'.bin',
+	'tsc'
+);
```
I recall some potential issues when trying to reference `node_modules` directly:

https://github.com/WordPress/gutenberg/pull/20215#discussion_r379176166

Not sure if it's something we expect to be problematic here, also given that this is referencing the top-level `node_modules/.bin` and not a package directory.

sirreal replied:
Resolving the package bin can be an involved process, I'd expect this to be reliable as long as we're on npm and the way packages are installed doesn't change radically. The failure should be obvious if and when it arises.

aduth commented on package.json:
```diff
@@ -188,14 +193,13 @@
 		"lint-pkg-json": "wp-scripts lint-pkg-json . 'packages/*/package.json'",
 		"lint-css": "wp-scripts lint-style '**/*.scss'",
 		"lint-css:fix": "npm run lint-css -- --fix",
-		"lint-types": "tsc",
```
From the original comment:

>Remove `lint-types` script, replaced by `build:packages-types`.

Just to confirm, will `build:packages-types` adequately fail builds where we're expecting type-checking to occur?

sirreal replied:
Yes, `npm run build:package-types` will error. It's part of `build:packages`. Should we add that to `lint`? That runs on the precommit hook?

aduth replied:
Yes, I think it would be good to have both as a pre-commit task, and reflected in the Travis build as well.

aduth commented on packages/a11y/tsconfig.json:
```diff
@@ -0,0 +1,10 @@
+	"compilerOptions": {
+		"rootDir": "src",
+		"declarationDir": "build-types",
+		"tsBuildInfoFile": "../../.cache/a11y.tsbuildinfo"
```
Is there an issue with using the default? Or conversely, is there value in having these all in the same top-level `.cache` directory?

sirreal replied:
The default is would be `.tsbuildinfo` in the package root (`packages/a11y/.tsbuildinfo`). That's probably fine to use.

One bit of nuance is that it seems these files must be removed to regenerate the type declaration files. That means cleaning `build-types` should be accompanied by cleaning `.tsbuildinfo`.

aduth commented on packages/a11y/tsconfig.json:
```diff
@@ -0,0 +1,10 @@
+		"declarationDir": "build-types",
+		"tsBuildInfoFile": "../../.cache/a11y.tsbuildinfo"
+	},
+	"references": [ { "path": "../dom-ready" } ],
```
I see some mention of project references in the original comment, but it's not really clear to me what purpose it serves, or how I as the implementing developer should know what to include and when. Is it something we could document?

sirreal replied:
In short, references should reflect any `@wordpress/*` (Gutenberg monorepo) dependencies for the package that also have types generated via `tsc`.

---

Do you have a good idea where it should go in the docs?

The TypeScript docs on [_Project References_ can be found here](https://www.typescriptlang.org/docs/handbook/project-references.html).

> Project references are a new feature in TypeScript 3.0 that allow you to structure your TypeScript programs into smaller pieces.
>
> By doing this, you can greatly improve build times, enforce logical separation between components, and organize your code in new and better ways.

This is analogous to packages the Gutenberg monorepo. The interesting bit is that `tsc` will take care of resolving an rebuilding referenced projects as required:

> Running tsc --build (tsc -b for short) will do the following:
> - Find all referenced projects
> - Detect if they are up-to-date
> - Build out-of-date projects in the correct order

aduth replied:
There's some prior art for documentation of packages development workflows at [`packages/README.md`](https://github.com/WordPress/gutenberg/blob/master/packages/README.md). Maybe a section there about TypeScript and/or `tsconfig.json` specifically?

sirreal replied:
I've [added documentation](https://github.com/WordPress/gutenberg/pull/18942/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.yml&short_path=353fb41#diff-353fb419a3a39a1b358ab0ff9cd560de).

aduth commented on bin/packages/build-types.js:
```diff
@@ -0,0 +1,38 @@
+
+main().catch( ( err ) => {
+	throw err;
+} );
```
Is the intention with this to account for Node warning about uncaught errors not exiting with non-zero status code in a future release? I wonder if that's any different if the `throw` takes place in `catch` (still part of the promise). Have you tested it?

Also, if we're handling _some_ errors in `main`, why not handle either all or no potential errors?

sirreal replied:
This isn't doing anything meaningful. I've made this all sync and removed try/catch to let errors surface.

sirreal committed:
No async, no catch

sirreal committed:
Use default .tsbuildinfo location

sirreal committed:
Clean up tsconfig

sirreal committed:
"Fix" token-list

relax strictness

sirreal committed:
Use eslint dis/enable no-console pair

sirreal committed:
Move types built to build:packages

sirreal commented on packages/i18n/src/index.js:
```diff
@@ -17,7 +17,10 @@ import sprintfjs from 'sprintf-js';
 	'': {
-		plural_forms: ( n ) => ( n === 1 ? 0 : 1 ),
+		/** @param {number} n */
+		plural_forms( n ) {
```
This fixes a local implicity any error, but the package still requires `noImplicitAny` because `tannin` and `memize` are untyped.

aduth replied:
> This fixes a local implicity any error, but the package still requires `noImplicitAny` because `tannin` and `memize` are untyped.

Sounds like a nice weekend project for me 😄 

(Previous WIP: https://github.com/aduth/tannin/tree/add/typescript-declarations)

aduth replied:
While I suppose [you'd already seen it](https://github.com/aduth/tannin/pull/4), just for posterity's sake, both Tannin and Memize now provide type definitions:

https://github.com/aduth/tannin/commit/b5f5fd958abae0d092baa6f867c660d98561bf2f
https://github.com/aduth/memize/commit/3a6f1cd38e25cf37716d2398c6ab098992acccfb

sirreal replied:
Yes, that's amazing! Thanks for adding those. We'd need `tannin@1.2.0` and `memize@1.1.0` it looks like. I won't take on upgrading those packages in Gutenberg myself right now, but when they land `i18n` should be able to drop the `noImplicityAny`.

sirreal replied:
Circled back to this: https://github.com/WordPress/gutenberg/pull/21224

sirreal commented on packages/token-list/tsconfig.json:
```diff
@@ -0,0 +1,10 @@
+		"rootDir": "src",
+		"outDir": "build-types",
+		"strictNullChecks": false,
+		"noImplicitAny": false
```
Undesireable, but there are lots of errors without it. This section is particularly problematic:

https://github.com/WordPress/gutenberg/blob/20186c99697f769ed0d2153335a491aeb91e477d/packages/token-list/src/index.js#L20-L22

It could likely be rewritten to enable more strict checks if we'd like.

aduth replied:
Is it any better if we simply "unwrap" it, [Prepack](https://prepack.io/)-style?

```js
this.entries = ( ...args ) => this._valueAsArray.entries( ...args );
this.forEach = ( ...args ) => this._valueAsArray.forEach( ...args );
this.keys = ( ...args ) => this._valueAsArray.keys( ...args );
this.values = ( ...args ) => this._valueAsArray.values( ...args );
```

sirreal replied:
I went ahead and got this strictly typed with some relatively minor restructuring in 106e3027d24c495cbeb695f9fab94c9a06944399

aduth replied:
Ohhhh, I remember this issue now. I ran into it in the past as well. I also recall it feeling like technically a bug in TypeScript, though admittedly our implementation is a pretty extreme edge case. Specifically, TypeScript doesn't detect the initialization of `_currentValue` and `_valueAsArray` in the constructor; they definitely are initialized there, but it's obscured by the fact that the initialization occurs as a side-effect of the setter associated with `this.value`.

sirreal committed:
Add TypeScript to packages/README

sirreal committed:
Ignore cli input

sirreal committed:
Add build:package-types to lint-staged

sirreal committed:
Add travis job

aduth commented:
From the changes in 75a1cf3 and 5598ee0 :

- Can we limit type-checking in precommit to files (or packages containing files) which are actually staged? 
   - Related: #18840
   - IIRC, `lint-staged` will pass files as arguments to the script
- It only just occurred to me we have separate entries for `*.js` and `packages/**/*.js` ([source](https://github.com/WordPress/gutenberg/blob/3ab37270ac91bec58ac6e137fd1d7d26b7f1b15d/package.json#L251-L254)), where I imagine this pre-commit step might be most relevant for the latter instead?
- Was it intentional to create a separate Travis job for this, vs. as it had been previously incorporated in the existing "Lint" job?

sirreal commented:
I'm hitting an issue with `is-shallow-equal` and I believe it's because the build output is not in a different directory tree from the source files. I've tried a number configurations for `tsconfig` with no success.

If the generated types exist and the TypeScript build needs to regenerated them, for example source changes or missing .tsbuildinfo, the job will error with:

> `TS5055: Cannot write file '…/build-types/….d.ts' because it would overwrite input file.`

It seems `packages/is-shallow-equals/build-types/*.d.ts` are being interpreted as source that cannot be overwritten. These are exactly the files we'd like to generate.

To reproduce this issue, do the following:

```sh
npm run clean:packages                                ## Clean everything (generated types and tsbuildinfo)
npm run build:package-types                           ## Generate declaration files
echo '// noop' >> packages/is-shallow-equal/index.js  ## Change file to trigger regeneration
npm run build:package-types                           ## !!! ERROR !!!
```

You'll see the following errors:

```
error TS5055: Cannot write file '…/gutenberg/packages/is-shallow-equal/build-types/arrays.d.ts' because it would overwrite input file.
error TS5055: Cannot write file '…/gutenberg/packages/is-shallow-equal/build-types/index.d.ts' because it would overwrite input file.
error TS5055: Cannot write file '…/gutenberg/packages/is-shallow-equal/build-types/objects.d.ts' because it would overwrite input file.
```

The same issue does not occur with other packages where the output directory is isolated from the source directors. For example, `a11y` does not have the same problem:

```sh
npm run clean:packages                        ## Clean everything (generated types and tsbuildinfo)
npm run build:package-types                   ## Generate declaration files
echo '// noop' >> packages/a11y/src/index.js  ## Change file to trigger regeneration
npm run build:package-types                   ## No problem
```

I've investigated (related issues linked below) and attempted many configuration combinations to get this to work as expected without success. I've gone as far as `packages/is-shallow-equals/tsconfig.json` as follows:

```jsonc
{
// We wouldn't use this configuration
// It demonstrates different suggestions that have been tried to fix the issue
	"compilerOptions": {
		"allowJs": true,
		"checkJs": true,
		"composite": true,
		"emitDeclarationOnly": true,
		"rootDir": ".",
		"outDir": "build-types",
		"declarationDir": "build-types"
	},
	"exclude": [ "build-types/**/*" ],
	"files": [ "arrays.js", "index.js", "objects.js" ],
	"include": [ "arrays.js", "index.js", "objects.js" ]
}
```

Apparently related issues:
- https://github.com/Microsoft/TypeScript/issues/16749
- https://github.com/microsoft/TypeScript/issues/14538
-  https://github.com/microsoft/TypeScript/issues/27436

sirreal committed:
Add and use lint-staged-typecheck bin

sirreal commented on bin/packages/lint-staged-typecheck.js:
```diff
@@ -0,0 +1,35 @@
+const tscPath = path.join( repoRoot, 'node_modules', '.bin', 'tsc' );
+
+// lint-staged passes full paths to staged changes
+/** @type {string[]} */
```
```suggestion
```

sirreal committed:
Update bin/packages/lint-staged-typecheck.js

sirreal commented:
> * Can we limit type-checking in precommit to files (or packages containing files) which are actually staged?
>   
>   * Related: #18840
>   * IIRC, `lint-staged` will pass files as arguments to the script

I've added a new bin for this and set it up with lint-staged. Here's an example:

![Screen Shot 2020-03-07 at 00 18 15](https://user-images.githubusercontent.com/841763/76130169-2383c880-600a-11ea-9c16-9dfab4bbee38.png)


> * It only just occurred to me we have separate entries for `*.js` and `packages/**/*.js` ([source](https://github.com/WordPress/gutenberg/blob/3ab37270ac91bec58ac6e137fd1d7d26b7f1b15d/package.json#L251-L254)), where I imagine this pre-commit step might be most relevant for the latter instead?

That's better, I've moved it to that section of lint-staged.

> * Was it intentional to create a separate Travis job for this, vs. as it had been previously incorporated in the existing "Lint" job?

That was intentional from my position of ignorance 🙂  I'm unfamiliar with the travis setup and didn't think about it, so I'm happy to put it wherever you suggest. It could be an additional script under the `Lint` job if you prefer. I generally differentiate between _lint_ and _type check_, but they are in the same category of code quality/static analysis tooling.

sirreal committed:
Restructure is-shallow-equal

sirreal commented:
> I'm hitting an issue with `is-shallow-equal` and I believe it's because the build output is not in a different directory tree from the source files. I've tried a number configurations for `tsconfig` with no success.

I've pushed a change in 4a0d7fa446e99aff11c89b7eaf4e2d7169d39f72 that restructures `is-shallow-equal` in a way that should be backwards compatible but works around the issue when generated type declarations should be overwritten.

aduth commented:
> That was intentional from my position of ignorance 🙂 I'm unfamiliar with the travis setup and didn't think about it, so I'm happy to put it wherever you suggest. It could be an additional script under the `Lint` job if you prefer. I generally differentiate between _lint_ and _type check_, but they are in the same category of code quality/static analysis tooling.

Yeah, I can see arguments either way. It also gets into the realm of pragmatic considerations like max concurrent builds in Travis, build performance (#15159), distributing tasks evenly based on total runtime, considering "reuse" of common setup (`npm ci`, etc).

I don't really feel too strongly about it, to be honest. [Current timings](https://travis-ci.com/github/WordPress/gutenberg/builds/152691171) are such that "Lint" is already at the higher end of the average. And we've got a fair bit of prior art for breaking these tasks down to be pretty granular. I think we can just keep it as proposed.

aduth commented:
> > I'm hitting an issue with `is-shallow-equal` and I believe it's because the build output is not in a different directory tree from the source files. I've tried a number configurations for `tsconfig` with no success.
> 
> I've pushed a change in [4a0d7fa](https://github.com/WordPress/gutenberg/commit/4a0d7fa446e99aff11c89b7eaf4e2d7169d39f72) that restructures `is-shallow-equal` in a way that should be backwards compatible but works around the issue when generated type declarations should be overwritten.

It's a conversation which has happened a few times in the past (https://github.com/WordPress/gutenberg/pull/13329#discussion_r264232869), but an implication of this change is that we are now building (transpiling) these files which we otherwise do not expect to be transpiled, [based on the current pattern for matching source files to build](https://github.com/WordPress/gutenberg/blob/a3aecaff89691250417ac4c3684c24ee2b5afea2/bin/packages/build.js#L129).

Likewise, this isn't going to be the only instance of this. Consider also [`eslint-plugin`](https://github.com/WordPress/gutenberg/tree/master/packages/eslint-plugin), [`npm-package-json-lint-config`](https://github.com/WordPress/gutenberg/tree/master/packages/npm-package-json-lint-config), etc (most commonly packages which are intended to be run in Node only).

I don't really have too much of an issue with consistently putting source files in a `src/` directory, but I think it needs to be a conscious and consistent shift of expectations, away from "the existence of `src/` is an indication of files which are built" to one of "source files are always in `src/`". This includes:

- Updating existing packages
- Updating the build script to use another indicator of files to be built

To be clear, I don't think these necessarily need to be done as part of this pull request, but I do think we should have consensus that:

- This is a sensible change we want to make
- There exists some other reasonable indicator we can use for files to be built

sirreal committed:
Bump to TS 3.8.3

sirreal committed:
Type TokenList

sirreal commented:
> […] an implication of this change is that we are now building (transpiling) these files which we otherwise do not expect to be transpiled, [based on the current pattern for matching source files to build](https://github.com/WordPress/gutenberg/blob/a3aecaff89691250417ac4c3684c24ee2b5afea2/bin/packages/build.js#L129) […]

Thanks for your thoughts and the background. I did not intend to start transpiling this package, that particular change was largely to validate what the fix was and start a conversation.

Source files in a project root often seem to be problematic when build systems are involved. One change with this PR is that these files were not previously interesting to a build system, but now they are.

There's a strong convention followed here that `src` contains uncompiled source files and `dist` contains compiled ouptut. I considered putting these files in `dist`, as they are intended for distribution, but that seemed likely to be more problematic, as compilation results are usually ignored and/or cleaned 🙂 

Another viable alternative would be a third directory containing "distributable source`, maybe `src-dist` or `dist-src`.
- `src` - to be compiled, not for use in distributed packages
- `dist` - compiled output, for use from published packages
- `dist-src` - uncompiled source for use from published packages

I don't have strong feelings regarding this or the options you outline. I'm happy if we can find any solution that will work with the changes in this PR.

sirreal committed:
Add project-management-automation to typed packages

sirreal committed:
Use lib for uncompiled published code

sirreal committed:
Fix implicit any type

sirreal commented:
- Rebased
- Updated to include `project-management-automation` in typed packages which had been added
- Changed `src` -> `lib` for uncompiled, published source based on (https://github.com/WordPress/gutenberg/pull/18942#issuecomment-601772189)

There's a new error from `project-management-automation` which is using stricter rules (`noImplicitAny`) now. Seems to be the same as this, which was fixed:  https://github.com/actions/toolkit/issues/199. Is the `@actions/github` dependency up to date?

sirreal committed:
Disable noImplicitAny for pma

sirreal commented:
A new error has come up we'll need to address:

```
packages/url/src/is-url.native.js:4:21 - error TS7016: Could not find a declaration file for module 'react-native-url-polyfill'. '/Users/jonsurrell/a8c/gutenberg/node_modules/react-native-url-polyfill/index.js' implicitly has an 'any' type.
```

sirreal committed:
Exclude react-native

sirreal commented:
> A new error has come up we'll need to address:
> 
> ```
> packages/url/src/is-url.native.js:4:21 - error TS7016: Could not find a declaration file for module 'react-native-url-polyfill'. '/Users/jonsurrell/a8c/gutenberg/node_modules/react-native-url-polyfill/index.js' implicitly has an 'any' type.
> ```

I've excluded `**/*.native.js` files. I suspect we'll need something like that long term, I don't know how other projects handle side-by-side web/native typings.

aduth commented:
>I've excluded `**/*.native.js` files. I suspect we'll need something like that long term, I don't know how other projects handle side-by-side web/native typings.

I have been pretty adamant in encouraging that public APIs (module exports) of native implementations would directly correspond to the equivalent web implementation (and vice-versa), so if the problem is that we can only allow one as the source of type detail, I think this is fine. 

Related:

- https://github.com/WordPress/gutenberg/pull/19918#issuecomment-580270379
- https://github.com/WordPress/gutenberg/pull/20231#issuecomment-586293328
- https://github.com/WordPress/gutenberg/pull/19947#issuecomment-585830875

sirreal commented:
> I have been pretty stubborn in encouraging that public APIs (module exports) of native implementations would directly correspond to the equivalent web implementation (and vice-versa)

I don’t know whether this will help. I’d expect `index.native.js` to look for type definitions as `index.native.d.ts`, so they might not get types for native-specific files.

sirreal committed:
Restore native checking, relax implicit any

aduth commented:
> I don’t know whether this will help. I’d expect `index.native.js` to look for type definitions as `index.native.d.ts`, so they might not get types for native-specific files.

Ah! Okay, that's a different problem than what I was thinking of, then. And I'm not too sure what the answer to that then is. I guess in a perfect world, the idea would be that `index.native.d.ts` would exist and simply point to `index.d.ts`? I say "perfect world" as in, I'm not sure whether it's technically feasible.

sirreal commented:
The issue was only with untyped `react-native-url-polyfill`. I've removed `**./*.native.js` from excludes and added set `noImplicitAny: false` for the `url` package.

This should reflect the existing setup. The relaxed any rule is undesirable but resolves this issue like we've done in several other packages already.

aduth commented on packages/is-shallow-equal/arrays/package.json:
```diff
@@ -0,0 +1,3 @@
+{
```
I suppose this is needed to support usage of `@wordpress/is-shallow-equal/arrays` ?

Does it work if this were `/index.js` of `module.exports = require( '../lib/arrays' );` ? Or does TypeScript take issue with that?

The reason I ask is two-fold:

- AFAIK, a `package.json` requires at minimum `name` and `version` fields to be valid [[1]](https://nodejs.org/en/knowledge/getting-started/npm/what-is-the-file-package-json/) [[2]](https://docs.npmjs.com/files/package.json). Whether the latter is relevant only for published packages, I'm neither sure nor feel confident one way or the other.
- It would be nice if we could avoid the `.npmpackagejsonlintrc.json`

sirreal replied:
I pushed a change with it. It seems to work fine with node and I'd expect with bundlers as well, however TypeScript (and IDE support) do not see to be able to find the types for imports like `@wordpress/is-shallow-equals/arrays`. I don't know if extra package.json with `types` fields will help, but I'm inclined not to provide first class support for this feature.

This is something that could be revisited, I'm curious about the motivations for allowing/supporting this.

aduth replied:
> This is something that could be revisited, I'm curious about the motivations for allowing/supporting this.

Was there anywhere you saw that we were referencing it this way? I vaguely recall something about `@wordpress/is-shallow-equal/array`, but at least in [how it's documented currently](https://github.com/WordPress/gutenberg/tree/master/packages/is-shallow-equal#usage), the object/array-specific functions are available as named exports. They're also available on the `wp` global as `wp.isShallowEqual.isShallowEqualArrays`, `wp.isShallowEqual.isShallowEqualObjects`. So I'm not sure it's something we considered as part of the public API, and was structured as such more to the point of general file structure?

sirreal replied:
Sounds like we can drop support for the direct imports via `…/arrays` and `…/objects`. That will help simplify things a little bit 👍

sirreal replied:
a3598cc removes the additional files to support direct access via `@wordpress/is-shallow-equal/{array,object}`.

aduth commented on packages/is-shallow-equal/.npmignore:
```diff
@@ -0,0 +1,2 @@
+benchmark
+test
```
I'd much rather have kept the whitelist and updated it per the new file structure, than to break convention (and [common recommendations](https://medium.com/@jdxcode/for-the-love-of-god-dont-use-npmignore-f93c08909d8d)) to resort to the blacklist.

sirreal replied:
4ed5b58 addresses this.

aduth commented on packages/project-management-automation/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+		"rootDir": "lib",
+		"declarationDir": "build-types",
+
+		// This is required due to a type error coming from missing types in @actions/github
```
🤔

sirreal replied:
Noted here in passing: https://github.com/WordPress/gutenberg/pull/18942#issuecomment-601793927

> There's a new error from `project-management-automation` which is using stricter rules (`noImplicitAny`) now. Seems to be the same as this, which was fixed: [actions/toolkit#199](https://github.com/actions/toolkit/issues/199). Is the `@actions/github` dependency up to date?

I haven't investigated and I'd like to leave it outside the scope here. `noImplicitAny` was disabled before which is why the error hadn't manifested.

sirreal committed:
Improve is-shallow-equal structure

- No package.json hack
- No package.json lint ignore
- Publish via files field

sirreal committed:
Proposal: Use root ts project

sirreal commented:
It would be nice to have someone familiar with Gutenberg Mobile test this out.

CC: @Tug @hypest Can you give this branch a test and see if this impacts your flows or help get a review from the mobile perspective?

sirreal commented on package.json:
```diff
@@ -252,7 +256,8 @@
 			"node ./bin/api-docs/update-readmes.js",
-			"node ./bin/api-docs/are-readmes-unstaged.js"
+			"node ./bin/api-docs/are-readmes-unstaged.js",
+			"node ./bin/packages/lint-staged-typecheck.js"
```
901ce5d07a01dc8a4691577f3f87e0ff796214d5 removed the custom build script and relies on a root `tsconfig.json` that references each project. This simplifies the type build significantly.

I'm inclined to have the `lint-staged` execute `npm run packages:build-types`. That would remove an additional custom script introduced here.

Once the TypeScript cache is warm, incremental builds are quick or instant. Running the entire type build has the benefit of detecting when type incompatibilities across packages are introduced, while the current version only detects that the _changed package itself_ typechecks and issues between packages would surface later on CI when the types are compiled.

@aduth Curious to hear your thoughts.

sirreal replied:
I attempted to do this and it's quite difficult because `lint-staged` always passes the changed files at the end, meaning we wind up with a command like the following:

```sh
npm run build:package-types "…/gutenberg/packages/a11y/src/index.js"

# tsc --build --verbose "…/gutenberg/packages/a11y/src/index.js"
# error TS6053: File '…/gutenberg/packages/a11y/src/index.js/tsconfig.json' not found.
```

If `lint-staged` is defined as a `js` file, [lint-staged tasks can be declared as functions](https://github.com/okonet/lint-staged#using-js-functions-to-customize-tasks) which makes it trivial to ignore the files input.

Are we open to that? I'd prefer to handle that in its own PR.

aduth replied:
> I'm inclined to have the `lint-staged` execute `npm run packages:build-types`. That would remove an additional custom script introduced here. [...] Once the TypeScript cache is warm, incremental builds are quick or instant.

Is the motivation to avoid maintaining the custom script, or because it may be faster to run a (warmed) full build than to typecheck individual package/files?

I'm motivated to keep these pre-commits fast. I'm not overly concerned with the maintenance burden of a custom script, if it means we can have faster/more granular checks.

>Running the entire type build has the benefit of detecting when type incompatibilities across packages are introduced, while the current version only detects that the changed package itself typechecks and issues between packages would surface later on CI when the types are compiled.

I agree it would be good to have more immediate feedback about potential issues. If it can be fast enough to not become a burden, then I like the idea. I'd also be content if at least this is considered in the full Travis build.

Aside: Slow pre-commit has become a pet-peeve of mine lately (see #18840), in case that helps contextualize my thoughts in approaching this.

sirreal replied:
> Is the motivation to avoid maintaining the custom script

This is a motivation

> or because it may be faster to run a (warmed) full build than to typecheck individual package/files?

A warm build of everything will likely be almost identical to a warm typecheck of changed files. Only the necessary things would be rebuilt.

The difference between a hot and cold cache is huge. The difference between a single package and all packages with a cold cache is also very large:

```sh
# Cold cache all
npm run clean:packages; time npm run build:package-types
# real    0m37.884s
# user    0m46.316s
# sys     0m1.311s

# Cold cache only URL (random package)
npm run clean:packages; time npm run build:package-types -- packages/url
# real    0m5.567s
# user    0m8.644s
# sys     0m0.353s

# Hot cache all
npm run build:package-types; time npm run build:package-types
# real    0m0.515s
# user    0m0.356s
# sys     0m0.091s
```

If a package has changed, it will be rebuilt. If it's dependencies have not been built, they will be built first. Given that rebuilding an unchanged package with a warm cache is practically free, the difference between building a single package and building all packages should be insignificant.

I agree these staged checks should be near immediate, and I don't feel too strongly about this.

We might try to persist TypeScript caches longer. TypeScript does an excellent job of doing absolutely nothing very quickly if it doesn't need to. I suspect it's only important to clean these types files before publish, so perhaps `tsc --build --clean` should be removed from `build:packages` and into the `publish:…` scripts and we could.

sirreal replied:
> We might try to persist TypeScript caches longer.

I've done this in 12302c0 where type cleaning is moved to publish scripts.

sirreal replied:
51195f349f5ef936b93b7c74de3addde0d623c8b proposes typechecking all modules on lint-staged. Feel free to play with it, but if you feel strongly we can roll this back. I don't feel strongly one way or the other.

sirreal commented:
Rebased to fix conflicts.

sirreal commented on bin/packages/lint-staged-typecheck.js:
```diff
@@ -0,0 +1,34 @@
+);
+
+try {
+	execa.sync( tscPath, [ '--build', ...changedPackages ] );
```
If we keep this script lint-staged, this could invoke `build:package-types` script with `-- ...changedPackages`.

I'd prefer to remove this entirely as noted: https://github.com/WordPress/gutenberg/pull/18942/files#r396317889

sirreal committed:
Use double quotes in package scripts

sirreal committed:
Only clean types on publish

sirreal committed:
Run lint-stage via npm script

sirreal committed:
Remove is-shallow-equal/{arrays,objects} support

sirreal committed:
Proposal: Check all types on lint-staged

sirreal committed:
Drop tsc verbose build

sirreal committed:
Fix types in token-list

aduth commented on bin/packages/lint-staged-typecheck.js:
```diff
@@ -0,0 +1,13 @@
+
+/* eslint-disable no-console */
+try {
+	execa.sync( 'npm', [ 'run', 'build:package-types' ] );
```
It might help to include a comment somewhere in here explaining why this file exists, vs. calling the NPM script directly from the pre-commit task. Otherwise, I fear it may not be obvious to a future maintainer that it need to continue to exist. Basically, what you had discovered at https://github.com/WordPress/gutenberg/pull/18942/files#r397735268 .

aduth reviewed (approved):
I appear to have given you a small bit more reconciliation work via #18840 😅 

As restitution, please accept my ✅

aduth commented:
This is going to be interesting for a little while, since switching from this branch to an older one will have quite a few "new" files.

One observation is that `.d.ts.map` files. Are those valuable / necessary? Should we be including in `package.json` `files` ?

sirreal commented:
> One observation is that `.d.ts.map` files. Are those valuable / necessary?

I've done some testing to see exactly how this changes things. For packages that include the declaration maps and original source, "go to definition" in a supporting IDE (e.g. VS Code), you'll be taken to the definition in source rather than the type declaration. For example:

```
import isShallowEqual from '@wordpress/is-shallow-equal';
```

Going to the definition of `isShallowEqual` here will:
- Go to `node_modules/@wordpress/is-shallow-equal/lib/index.js` with declaration map
- Go to `node_modules/@wordpress/is-shallow-equal/build-types/index.d.ts` without declaration map

Documented here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-9.html#new---declarationmap

> Should we be including in `package.json` `files` ?

They should be included because they're under `build-types`, which should be included via `files`.

You might have some stale files sitting around from intermediate versions of these changes. All `.d.ts` and `.d.ts.map` files should not be under `build-types`. Before moving `is-shallow-equal` source into `lib`, these files would be generated next to the source in the package root and are not ignored.

sirreal committed:
Explain lint-staged script with a comment

sirreal committed:
Restore bin typechecking

sirreal committed:
Comment cache outdir

sirreal commented:
Rebase to fix conflicts and incorporate `bin` typechecking into the new composite project setup.

I think this is ready to go! 🚀 

@aduth would you mind giving this some final testing after the rebase?

sirreal commented on packages/is-shallow-equal/package.json:
```diff
@@ -20,11 +20,13 @@
-		"objects.js"
+		"lib",
+		"build-types",
+		"*.js",
```
```suggestion
```

sirreal committed:
Remove redundant entry from is-shallow-equal files

sirreal committed:
Disable declarationMap for bin

gziolo commented on packages/is-shallow-equal/package.json:
```diff
@@ -20,11 +20,12 @@
-		"arrays.js",
-		"index.js",
-		"objects.js"
+		"lib",
```
@aduth, isn’t it a breaking change to rename files? I recall you could use file imports to use low level functions. Was it documented or I made that up?

I’m not against the change, I like it but we should add CHANGELOG entry if it’s confirmed.

aduth replied:
Related previous discussion: https://github.com/WordPress/gutenberg/pull/18942#discussion_r395842392

Do you recall where said usage was promoted? I have vague memory of it. The low-level functions are still available as named exports.

gziolo replied:
Before this PR: https://github.com/WordPress/gutenberg/pull/12825
It’s safe to proceed but we need to update benchmarks:
https://github.com/WordPress/gutenberg/blob/a2884c129598a2f020757820a7e52a3d2920e596/packages/is-shallow-equal/benchmark/index.js#L19-20

sirreal replied:
Nice catch, thanks! Addressed:

- 6ee66b50f8 Update is-shallow-equal tests
- a213ca951c Update is-shallow-equal benchmarks

gziolo commented on packages/a11y/package.json:
```diff
@@ -21,6 +21,7 @@
 	"main": "build/index.js",
 	"module": "build-module/index.js",
 	"react-native": "src/index",
+	"types": "build-types",
```
Shouldn’t it be listed in the linter for package.json files in the rule that validates the order of fileds?

sirreal replied:
👍 Added in adb7a78c808f29154e5e356523b8bd0dc9a9d775

gziolo commented on package.json:
```diff
@@ -167,16 +170,18 @@
 		"prebuild": "npm run check-engines",
-		"clean:packages": "rimraf ./packages/*/build ./packages/*/build-module ./packages/*/build-style ./packages/*/node_modules",
+		"clean:packages": "rimraf \"./packages/*/@(build|build-module|build-style)\"",
+		"clean:package-types": "tsc --build --clean",
```
Shouldn’t it be part of clean:packages that is executed as prebuild:packages?

sirreal replied:
There's some conversation around this in the thread here: https://github.com/WordPress/gutenberg/pull/18942#discussion_r396317889

In summary, there's enormous benefit to having a warm TypeScript cache and I don't believe there's any reason to clean it except to remove outdated declaration files from packages before they will be published.

In particular, we'll want to keep the caches warm for typechecking as part of lint staged.

sirreal committed:
Add types to package.json lint

sirreal committed:
Update is-shallow-equal benchmarks

sirreal committed:
Update is-shallow-equal tests

sirreal committed:
Revert "Explain lint-staged script with a comment"

This reverts commit 8dffe92dbd4abd3f0469de844239e5ae3a1c51d7.

sirreal committed:
Revert "Proposal: Check all types on lint-staged"

This reverts commit 70b1c2e21e99caf7cdd7175e646eb6fc9ce22e5f.

sirreal committed:
Revert "Run lint-stage via npm script"

This reverts commit e2d8f8675f22b8dc8c3c4fe897581db626f4331f.

sirreal committed:
Log type errors with lint-staged

sirreal commented:
All feedback has been addressed.

I've done some final tweaking related to `lint-staged`:

#### Only typecheck changed packages

See this thread for some details: https://github.com/WordPress/gutenberg/pull/18942#discussion_r396317889

I do believe there's value in typechecking everything, but initial builds can be greater than thirty seconds. I'd prefer to avoid introducing significant overhead abruptly. We can consider typechecking all packages on lint-staged in the future.

#### Ensure lint-staged prints an actionable error.

Somewhere between execa version changes and usage of the `packages:build-types` npm script, actionable output from `tsc` errors was lost. Invoke `tsc` directly and capture/print error messages.

Example:

```none
$ git commit

husky > pre-commit (node v12.15.0)
  ↓ Stashing changes... [skipped]
    → No partially staged files found...
  ❯ Running tasks...
    ↓ Running tasks for packages/*/package.json [skipped]
      → No staged files match packages/*/package.json
    ↓ Running tasks for *.scss [skipped]
      → No staged files match *.scss
    ✔ Running tasks for *.js
    ↓ Running tasks for {docs/{toc.json,tool/*.js},packages/{*/README.md,components/src/*/**/README.md}} [skipped]
      → No staged files match {docs/{toc.json,tool/*.js},packages/{*/README.md,components/src/*/**/README.md}}
    ❯ Running tasks for packages/**/*.js
      ✔ node ./bin/api-docs/update-api-docs.js
      ✔ node ./bin/api-docs/are-api-docs-unstaged.js
      ✖ node ./bin/packages/lint-staged-typecheck.js



✖ node ./bin/packages/lint-staged-typecheck.js found some errors. Please fix them and try committing again.

packages/a11y/src/index.js(24,17): error TS2345: Argument of type '11' is not assignable to parameter of type 'string'.
```

sirreal committed:
Update changelog for affected packages

- New feature for packages with type declarations
- Breaking change for is-shallow-equal direct imports of arrays/objects
- Breaking change for `types` order in npm-package-json-lint-config

sirreal commented:
I've updated changelogs, see dfd2bc1969f0e8e29cfea28c6d67d97ae7500686 for details:

> Update changelog for affected packages
> - New feature for packages with type declarations
> - Breaking change for is-shallow-equal direct imports of arrays/objects
> - Breaking change for `types` order in npm-package-json-lint-config

Tug commented:
I did a quick sanity check and it looks good for gutenberg-mobile as of now 👍 

It might impact the Monorepo work that's being done though as we needed to do some changes in `tsconfig.json` to fix a conflict with react-native types. I guess we can address that ourselves though and ping you if needed ;)

gziolo commented:
Travis is green, there is approval from @aduth. I will leave it to you both to push the button and celebrate this moment. Awesome work 🥇 

I'm looking forward to starting using it in practice – I wasn't very happy by type support in Visual Studio Code for code imported from local npm packages. It should make it way much better! :)

aduth commented:
> I wasn't very happy by type support in Visual Studio Code for code imported from local npm packages. It should make it way much better! :)

Initially it will be better in some ways (up-to-date) and worse in others (our usage of types currently is not very good), but it will definitely force us to be more honest with how and where we use types. I imagine it will only get better over time! 🚀

aduth reviewed (approved):
Still looking good 👍

sirreal commented:
Thanks everyone who helped get this over the line ❤️ 🙌
